src/library/export/coverartcopyworker.cpp

```diff
@@ -0,0 +1,98 @@
+#include "library/export/coverartcopyworker.h"
+
+#include <QDebug>
+#include <QFileInfo>
+#include <QMessageBox>
+
+#include "util/fileaccess.h"
+#include "util/imagefiledata.h"
+#include "util/safelywritablefile.h"
+
+void CoverArtCopyWorker::run() {
+    auto selectedCover = mixxx::FileAccess(mixxx::FileInfo(m_selectedCoverArtFilePath));
+    ImageFileData imageFileData = ImageFileData::fromFilePath(m_selectedCoverArtFilePath);
+    if (imageFileData.isNull()) {
+        // TODO(rryan): feedback
+        return;
+    }
+
+    m_coverInfo.type = CoverInfo::FILE;
+    m_coverInfo.source = CoverInfo::USER_SELECTED;
+    m_coverInfo.coverLocation = m_selectedCoverArtFilePath;
+    m_coverInfo.setImage(imageFileData);
+
+    if (QFileInfo(m_oldCoverArtFilePath).canonicalPath() ==
+            QFileInfo(m_selectedCoverArtFilePath).canonicalPath()) {
+        qDebug() << "Track and selected cover art are in the same path:"
+                 << QFileInfo(m_selectedCoverArtFilePath).canonicalPath()
+                 << "Cover art updated without copying";
+        emit updateCoverArt(m_coverInfo);
+        return;
+    }
+
+    copyFile(m_selectedCoverArtFilePath, m_oldCoverArtFilePath);
+}
+
+void CoverArtCopyWorker::copyFile(
+        const QString& m_selectedCoverArtFilePath,
+        const QString& m_oldCoverArtFilePath) {
+    QFileInfo coverArtPathFileInfo(m_oldCoverArtFilePath);
+    ImageFileData imageFileData = ImageFileData::fromFilePath(m_selectedCoverArtFilePath);
+
+    if (coverArtPathFileInfo.exists()) {
+        switch (makeOverwriteRequest(m_oldCoverArtFilePath)) {
+        case OverwriteAnswer::Cancel:
+            return;
+        case OverwriteAnswer::Overwrite:
+            break;
+        }
+
+        mixxx::SafelyWritableFile safelyWritableFile(m_oldCoverArtFilePath,
+                mixxx::SafelyWritableFile::SafetyMode::Replace);
+
+        DEBUG_ASSERT(!safelyWritableFile.fileName().isEmpty());
+        if (imageFileData.saveFile(safelyWritableFile.fileName())) {
+            qDebug() << "Cover art"
+                     << m_oldCoverArtFilePath
+                     << "copied successfully";
+            safelyWritableFile.commit();
+        } else {
+            qWarning() << "Error while copying the cover art to" << safelyWritableFile.fileName();
+        }
+    } else {
+        if (imageFileData.saveFile(m_oldCoverArtFilePath)) {
+            qDebug() << "Cover art"
+                     << m_oldCoverArtFilePath
+                     << "copied successfully";
+        } else {
+            qWarning() << "Error while copying the cover art to" << m_oldCoverArtFilePath;
+        }
+    }
+    emit updateCoverArt(m_coverInfo);
+    quit();
```

Comment:
This is redundant, because the thread ends after run() anyway. It is also misplaced in the sub function.